esmero/format_strawberryfield
# ISSUE-22: Implement Logic for exposed Metadata Endpoints

# What is new?

See #22 

What this pull adds the missing logic to our dynamic Metadata Endpoint Controller that acts on ADOs using Metadata Display Entities as processors. There is a lot to be done still on this but i prefer to move slowly and get functionality in and then refine.

# How does this work ?

- We expose a dynamic route at
'/node/{node}/metadata/{metadataexposeconfig_entity}/{format}'
Where {node} is the numeric id of a node
Is the machine name of a configurable metadataexposeconfig_entity, which allows people to bind an existing Metadata Display (Twig) Entity to the configuration, and a bundle set restriction
and {format}, not used yet, and to be configurable stub filename so we can match some APIs expectations (like manifest.json)
- On a call to this route, we load the node, check if the type is allowed, check there are SBF fields there (redundant since we really only allow bundles with SBF but better to be safe), read the configured metadata display, do some logic calling its own renderNative method encapsulated into a Render Context to fetch any Cacheable Metadata that could be bubble up from the Twig invocation, we get the set mimetype response and generate a cacheable Response that respects all those things + adds each involved Cacheable object in the response. 

This last means that if a Metadata Display Entity (twig) does not change and the ADO/Node does not change we will returning cached values, blazing fast.

We also Throw some HTTP exceptions if something goes wrong. That needs some research since we would probably want to make sure that the responses are in the same mimetype/content-type header format as the correct one to make sure any API calling us has a better time dealing with us.

- The metadata Display entity also got way smarter by dealing with its own rendering methods and Defined Variables Parsing. Specially Proud of https://github.com/esmero/format_strawberryfield/blob/ISSUE-22/src/Entity/MetadataDisplayEntity.php#L392-L440 which will allow us to assist users in writing and using this type of entiti8es and generate even sample data for each Twig template.

# How to test?

Setup a new Metadata Expose Endpoint at (Configure Metadata Exposed endpoints at) http://localhost:8001/admin/config/archipelago/metadataexpose

Name it "iiifmanifest", the label will match the machine name but if you use a larger label with spaces machine name can be different. You can also edit the machine name. The machine name is what gets used in the path.
Choose ¨IIIF Presentation API 3.0 Manifest¨as the `The Metadata display Entity (Twig) to be used to generate data at this endpoint`

Check Digital Object as Which Content types will this metadata be allowed to be exposed?
and make sure "Is this exposed Metadata Endpoint active?" is checked. Note: i still need to make sure we actually check if its active or not... or did i do that already??

Save.

Now go to some node path using the given format, screenshot down here is more explicit

![metadata-expose-output](https://user-images.githubusercontent.com/6946023/67324521-ecd96580-f4e1-11e9-80b6-b22843f78023.jpg)


Now test the same with another endpoint that uses, e.g the MODS 3.6 Metadata Display. 

Try to access nodes that are not of the given type, etc, iterate and hopefully break so i can make this more robust.

Also do a "curl" to check if content-type matches. There will be extra work on figuring out extra headers here.


@marlo-longley @giancarlobi if any of you mind looking at it. I will probably push more code to address thing i know need more work, like adding extra config options, an extra button to test nodes of type "X" randomly during the setup and being explicit ("just some text") where people can access the path when dealing with this and how to use also this endpoints inside other Twig templates! (real linked data folks!)
